File: sdk/python/touca/_options.py
# Copyright 2023 Touca, Inc. Subject to Apache-2.0 License.

from argparse import Action, ArgumentParser
from configparser import ConfigParser
from json import loads
from pathlib import Path
from typing import Any, Dict, List

from touca._transport import Transport


class ToucaError(RuntimeError):
    errors = {
        "auth_invalid_key": {
            "code": "E01",
            "message": "Authentication failed: API Key Invalid.",
        },
        "auth_invalid_response": {
            "code": "E02",
            "message": "Authentication failed: Invalid Response.",
        },
        "auth_server_down": {
            "code": "E03",
            "message": "Touca server appears to be down",
        },
        "config_file_missing": {
            "code": "E04",
            "message": 'Configuration file does not exist: "{}"',
        },
        "config_file_invalid": {
            "code": "E05",
            "message": 'Configuration file has an unexpected format: "{}"',
        },
        "config_option_invalid": {
            "code": "E06",
            "message": 'Configuration option "{}" has unexpected type.',
        },
        "config_option_missing": {
            "code": "E07",
            "message": 'Configuration option "{}" is missing.',
        },
        "config_option_sealed": {
            "code": "E08",
            "message": "The specified version is already submitted.",
        },
        "config_option_fetch": {
            "code": "E09",
            "message": "Failed to fetch options from the remote server.{}",
        },
        "config_workflows_missing": {
            "code": "E10",
            "message": "No workflow is registered.",
        },
        "capture_not_configured": {
            "code": "E11",
            "message": "Client not configured to perform this operation.",
        },
        "capture_forget": {
            "code": "E12",
            "message": 'Test case "{}" was never declared.',
        },
        "capture_type_mismatch": {
            "code": "E13",
            "message": 'Specified key "{}" has a different type.',
        },
        "transport_http": {
            "code": "E14",
            "message": "HTTP request failed: {}",
        },
        "transport_post": {
            "code": "E15",
            "message": "Failed to submit test results.{}",
        },
        "transport_seal": {
            "code": "E16",
            "message": "Failed to seal this version.",
        },
    }

    def __init__(self, key: str, *args):
        entry = ToucaError.errors[key]
        self.key = key
        self.link = f"https://touca.io/docs/sdk/errors#{entry['code'].lower()}-{key}"
        super().__init__(entry["message"].format(*args))


def find_home_path():
    path = Path.cwd().joinpath(".touca")
    return path if path.exists() else Path.home().joinpath(".touca")


def find_profile_path():
    home_path = find_home_path()
    settings_path = Path(home_path, "settings")
    name = "default"
    if settings_path.exists():
        config = ConfigParser()
        config.read_string(settings_path.read_text())
        name = config.get("settings", "profile")
    return home_path.joinpath("profiles", name)


def parse_config_profile():
    path = find_profile_path()
    if path.exists():
        config = ConfigParser()
        config.read_string(path.read_text())
        return config


def throw_if_missing(options, keys):
    missing = [key for key in keys if key not in options]
    if missing:
        raise ToucaError("config_option_missing", missing[0])


def validate_options_type(options: dict, cls, keys: List[str]):
    invalid = [
        key for key in keys if key in options and not isinstance(options[key], cls)
    ]
    if invalid:
        raise ToucaError("config_option_invalid", invalid[0])


def fixup_flags(options: dict):
    for k in [
        "colored_output",
        "offline",
        "overwrite_results",
        "save_binary",
        "save_json",
    ]:
        if k in options:
            options[k] = options.get(k) in [True, "True", "true"]


def prepare_parser(parser: ArgumentParser):
    class ExtendAction(Action):
        def __call__(self, parser, namespace, values, option_string=None):
            items = getattr(namespace, self.dest) or []
            items.extend(values)
            setattr(namespace, self.dest, items)

    parser.register("action", "extend", ExtendAction)
    parser.add_argument("--api-key", help="Touca API Key", dest="api_key")
    parser.add_argument("--api-url", help="Touca API URL", dest="api_url")
    parser.add_argument(
        "--team",
        dest="team",
        help="Slug of team to which test results belong",
    )
    parser.add_argument(
        "--suite",
        dest="suite",
        help="Slug of suite to which test results belong",
    )
    parser.add_argument(
        "--revision",
        dest="version",
        help="Version of the code under test",
    )
    parser.add_argument(
        "--offline",
        dest="offline",
        const=True,
        default=False,
        nargs="?",
        help="Disables all communications with the Touca server",
    )
    parser.add_argument(
        "--save-as-binary",
        dest="save_binary",
        const=True,
        default=False,
        nargs="?",
        help="Save a copy of test results on local filesystem in binary format",
    )
    parser.add_argument(
        "--save-as-json",
        dest="save_json",
        const=True,
        default=False,
        nargs="?",
        help="Save a copy of test results on local filesystem in JSON format",
    )
    parser.add_argument(
        "--output-directory",
        dest="output_directory",
        help="Path to a local directory to store result files",
    )
    parser.add_argument(
        "--overwrite",
        dest="overwrite_results",
        const=True,
        default=False,
        nargs="?",
        help="Overwrite result directory for testcase if it already exists",
    )
    parser.add_argument(
        "--testcase",
        "--testcases",
        dest="testcases",
        action="extend",
        nargs="+",
        help="One or more testcases to feed to the workflow",
    )
    parser.add_argument(
        "--filter",
        dest="workflow_filter",
        help="Name of the workflow to run",
    )
    parser.add_argument(
        "--log-level",
        dest="log_level",
        choices=["debug", "info", "warn"],
        default="info",
        help="Level of detail with which events are logged",
    )
    parser.add_argument(
        "--colored-output",
        dest="colored_output",
        const=True,
        default=True,
        nargs="?",
        help="Use color in standard output",
    )
    parser.add_argument(
        "--config-file",
        help="Path to a configuration file",
        dest="config_file",
    )


def assign_options(target: dict, source: dict):
    target_keys = {
        "api_key": "api_key",
        "api_url": "api_url",
        "team": "team",
        "suite": "suite",
        "version": "version",
        "offline": "offline",
        "save_binary": "save_binary",
        "save_json": "save_json",
        "output_directory": "output_directory",
        "overwrite_results": "overwrite_results",
        "testcases": "testcases",
        "workflow_filter": "workflow_filter",
        "log_level": "log_level",
        "colored_output": "colored_output",
        "config_file": "config_file",
        "api-key": "api_key",
        "api-url": "api_url",
        "revision": "version",
        "save-as-binary": "save_binary",
        "save-as-json": "save_json",
        "output-directory": "output_directory",
        "overwrite": "overwrite_results",
        "filter": "workflow_filter",
        "log-level": "log_level",
        "colored-output": "colored_output",
        "config-file": "config_file",
        "submit_async": "submit_async",
    }
    for key, value in source.items():
        if value is not None and key in target_keys:
            target[target_keys[key]] = value


def apply_environment_variables(options):
    from os import environ

    assign_options(
        options,
        {
            "api_key": environ.get("TOUCA_API_KEY"),
            "api_url": environ.get("TOUCA_API_URL"),
            "version": environ.get("TOUCA_TEST_VERSION"),
        },
    )


def apply_cli_arguments(options: dict):
    from sys import argv

    parser = ArgumentParser(
        description="Touca Test Runner",
        epilog="See https://touca.io/docs for more information.",
    )
    prepare_parser(parser)
    parsed = vars(parser.parse_known_args(options.get("arguments", argv[1:]))[0])
    assign_options(options, parsed)
    fixup_flags(options)


def apply_config_file(options: dict):
    if "config_file" not in options:
        return
    file = Path(options["config_file"]).resolve()
    if not file.is_file():
        raise ToucaError("config_file_missing", file)
    try:
        parsed = loads(file.read_text())
    except ValueError:
        raise ToucaError("config_file_invalid", file)
    if "touca" not in parsed:
        raise ToucaError("config_file_invalid", file)
    assign_options(options, parsed.get("touca"))


def apply_config_profile(options: dict):
    config = parse_config_profile()
    if config and config.has_section("settings"):
        assign_options(options, config["settings"])
        fixup_flags(options)


def apply_api_url(options: Dict[str, str]):
    from urllib.parse import urlparse

    if "api_url" not in options:
        return
    url = urlparse(options["api_url"])
    url_path: List[str] = [k.strip("/") for k in url.path.split("/@/")]
    options["api_url"] = f"{url.scheme}://{url.netloc}/{url_path[0]}".rstrip("/")
    if len(url_path) == 1:
        return
    slugs = [k for k in url_path[1].split("/") if k]
    assign_options(options, dict(zip(["team", "suite", "version"], slugs)))


def apply_core_options(options: dict):
    options.setdefault("concurrency", True)
    if not options.get("offline"):
        options["offline"] = all(x not in options for x in ["api_key", "api_url"])
    if "api_key" in options and "api_url" not in options:
        options["api_url"] = "https://api.touca.io"


def authenticate(options: dict, transport: Transport):
    if (
        options.get("offline") is False
        and "api_key" in options
        and "api_url" in options
    ):
        transport.configure(options)


def apply_server_options(options: dict, transport: Transport):
    if options.get("offline") or "api_url" not in options:
        return
    response = transport.request(method="GET", path="/platform")
    if response.status != 200:
        raise ToucaError("auth_server_down")
    body: dict = loads(response.data.decode("utf-8"))
    if not body.get("ready"):
        raise ToucaError("auth_server_down")
    if "webapp" in body:
        options["web_url"] = body.get("webapp", None)


def apply_runner_options(options: dict):
    options.setdefault("submit_async", False)
    options.setdefault("output_directory", find_home_path().joinpath("results"))
    options.setdefault("workflows", [])
    if "workflow_filter" in options:
        options["workflows"] = list(
            filter(
                lambda x: x.get("suite") == options["workflow_filter"],
                options["workflows"],
            )
        )
        del options["workflow_filter"]
    for v in options["workflows"]:
        assign_options(
            v,
            {
                "suite": options.get("suite"),
                "version": options.get("version"),
                "testcases": options.get("testcases"),
            },
        )
    options.pop("suite", None)
    options.pop("version", None)
    options.pop("testcases", None)


def fetch_remote_options(input, transport: Transport):
    response = transport.request(method="POST", path="/client/options", body=input)
    if response.status == 401:
        raise ToucaError("auth_invalid_key")
    if response.status == 409:
        raise ToucaError("config_option_sealed")
    if response.status != 200:
        reason = ""
        if response.status == 404:
            reason = " This team does not exist." 
        if response.status == 409:
            reason = " This version is sealed."
        raise ToucaError("config_option_fetch", reason)
    return loads(response.data.decode("utf-8"))


def apply_remote_options(options: dict, transport: Transport):
    if options.get("offline") or "api_key" not in options or "api_url" not in options:
        return
    workflows: List[Dict[str, Any]] = options["workflows"]
    if not workflows:
        return
    response = fetch_remote_options(
        [
            {
                "team": options.get("team"),
                "suite": v.get("suite"),
                "version": v.get("version"),
                "testcases": v.get("testcases", []),
            }
            for v in workflows
        ],
        transport,
    )
    for v in response:
        w = next(x for x in workflows if x.get("suite") == v.get("suite"))
        if w:
            assign_options(
                w, {"version": v.get("version"), "testcases": v.get("testcases")}
            )


def validate_core_options(options: dict):
    validate_options_type(options, bool, ["concurrency", "offline"])
    validate_options_type(
        options, str, ["api_key", "api_url", "suite", "team", "version"]
    )
    slugs = ["team", "suite", "version"]
    if any(x in options for x in slugs):
        throw_if_missing(options, slugs)
    if not options.get("offline"):
        throw_if_missing(options, ["api_key", "api_url"])
    return (
        False
        if any(options.get(x) for x in slugs) and not all(options.get(x) for x in slugs)
        else True
        if options.get("offline")
        else all(x in options for x in ["api_key", "api_url"])
    )


def validate_runner_options(options: dict):
    validate_options_type(
        options,
        bool,
        [
            "colored_output",
            "concurrency",
            "offline",
            "overwrite_results",
            "save_binary",
            "save_json",
        ],
    )
    validate_options_type(
        options,
        str,
        [
            "api_key",
            "api_url",
            "suite",
            "team",
            "version",
            "workflow_filter",
        ],
    )
    validate_options_type(options, Path, ["output_directory"])
    if not options.get("offline"):
        throw_if_missing(options, ["api_key", "api_url"])
    workflows = options.get("workflows")
    if not workflows:
        raise ToucaError("config_workflows_missing")
    if not all(w.get("version") for w in workflows):
        raise ToucaError("config_option_missing", "version")
    if not all(w.get("testcases") for w in workflows):
        raise ToucaError("config_option_missing", "testcases")
    if options.get("offline"):
        throw_if_missing(options, ["team"])
        for w in workflows:
            throw_if_missing(w, ["suite", "version"])


def update_core_options(options: dict, transport: Transport):
    apply_environment_variables(options)
    apply_api_url(options)
    apply_core_options(options)
    authenticate(options, transport)
    return validate_core_options(options)


def update_runner_options(options: dict, transport: Transport):
    apply_cli_arguments(options)
    apply_config_file(options)
    apply_config_profile(options)
    apply_environment_variables(options)
    apply_api_url(options)
    apply_core_options(options)
    authenticate(options, transport)
    apply_server_options(options, transport)
    apply_runner_options(options)
    apply_remote_options(options, transport)
    validate_runner_options(options)

File: sdk/cpp/src/runner/options.cpp
// Copyright 2022 Touca, Inc. Subject to Apache-2.0 License.

#include "touca/runner/detail/options.hpp"

#include "cxxopts.hpp"
#include "rapidjson/document.h"
#include "touca/client/detail/options.hpp"
#include "touca/core/filesystem.hpp"
#include "touca/devkit/platform.hpp"
#include "touca/devkit/utils.hpp"

namespace touca {
namespace detail {

cxxopts::Options cli_options() {
  cxxopts::Options options("./app", "Command Line Options");

  // clang-format off
  options.add_options("main")
      ("h,help", "displays this help message")
      ("v,version", "prints version of this executable")
      ("c,config-file",
          "path to configuration file",
          cxxopts::value<std::string>())
      ("api-key",
          "Touca server api key",
          cxxopts::value<std::string>())
      ("api-url",
          "Touca server api url",
          cxxopts::value<std::string>())
      ("team",
          "slug of team to which testresults belong",
          cxxopts::value<std::string>())
      ("suite",
          "slug of suite to which testresults belong",
          cxxopts::value<std::string>())
      ("r,revision",
          "version to associate with testresults",
          cxxopts::value<std::string>())
      ("testcase",
          "one or more testcases to feed to the workflow",
          cxxopts::value<std::vector<std::string>>())
      ("testcase-file",
          "single file listing testcases to feed to the workflow",
          cxxopts::value<std::string>())
      ("o,output-dir",
          "path to output directory",
          cxxopts::value<std::string>()->default_value("./results"))
      ("log-level",
          "level of detail with which events are logged",
          cxxopts::value<std::string>()->default_value("info"))
      ("save-as-binary",
          "save a copy of test results on local disk in binary format",
          cxxopts::value<bool>()->implicit_value("true")->default_value("true"))
      ("save-as-json",
          "save a copy of test results on local disk in json format",
          cxxopts::value<bool>()->implicit_value("true")->default_value("false"))
      ("skip-logs",
          "do not generate log files",
          cxxopts::value<bool>()->implicit_value("true"))
      ("redirect-output",
          "redirect content printed to standard streams to files",
          cxxopts::value<bool>()->default_value("true"))
      ("offline",
          "do not submit results to Touca server",
          cxxopts::value<bool>()->implicit_value("true"))
      ("overwrite",
          "overwrite result directory for testcase if it already exists",
          cxxopts::value<bool>()->implicit_value("true"))
      ("colored-output",
          "use color in standard output",
          cxxopts::value<bool>()->default_value("true"));
  // clang-format on

  return options;
}

template <typename T>
void parse_cli_option(const cxxopts::ParseResult& result,
                      const std::string& key, T& field) {
  if (result[key].count()) {
    field = result[key].as<T>();
  }
}

void parse_file_option(const rapidjson::Value& result, const std::string& key,
                       std::string& field) {
  if (result.HasMember(key) && result[key].IsString()) {
    field = result[key].GetString();
  }
}

void parse_file_option(const rapidjson::Value& result, const std::string& key,
                       bool& field) {
  if (result.HasMember(key) && result[key].IsBool()) {
    field = result[key].GetBool();
  }
}

/**
 * @param argc number of arguments provided to the application
 * @param argv list of arguments provided to the application
 * @param options application configuration parameters
 */
bool parse_cli_options(int argc, char* argv[], FrameworkOptions& options) {
  auto opts = cli_options();
  opts.allow_unrecognised_options();
  try {
    const auto& result = opts.parse(argc, argv);
    if (result.count("help")) {
      options.has_help = true;
    }
    if (result.count("version")) {
      options.has_version = true;
    }
    if (result.count("testcase")) {
      options.testcases = result["testcase"].as<std::vector<std::string>>();
    }
    options.output_dir = result["output-dir"].as<std::string>();
    options.log_level = result["log-level"].as<std::string>();
    options.save_binary = result["save-as-binary"].as<bool>();
    options.save_json = result["save-as-json"].as<bool>();
    options.redirect = result["redirect-output"].as<bool>();
    options.colored_output = result["colored-output"].as<bool>();
    parse_cli_option(result, "api-key", options.api_key);
    parse_cli_option(result, "api-url", options.api_url);
    parse_cli_option(result, "config-file", options.config_file);
    parse_cli_option(result, "testcase-file", options.testcase_file);
    parse_cli_option(result, "team", options.team);
    parse_cli_option(result, "suite", options.suite);
    parse_cli_option(result, "revision", options.revision);
    parse_cli_option(result, "skip-logs", options.skip_logs);
    parse_cli_option(result, "offline", options.offline);
    parse_cli_option(result, "overwrite", options.overwrite);
  } catch (const cxxopts::OptionParseException& ex) {
    touca::print_error("failed to parse command line arguments: {}\n",
                       ex.what());
    return false;
  }

  return true;
}

bool parse_file_options(FrameworkOptions& options) {
  // if user is asking for help description or framework version,
  // do not parse the configuration file even if it is specified.

  if (options.has_help || options.has_version) {
    return true;
  }

  // if configuration file is not specified yet a file config.json
  // exists in current directory, attempt to use that file.

  if (options.config_file.empty()) {
    if (!touca::filesystem::is_regular_file("./config.json")) {
      return true;
    }
    options.config_file = "./config.json";
  }

  // configuration file must exist if it is specified

  if (!touca::filesystem::is_regular_file(options.config_file)) {
    touca::print_error("configuration file not found: {}\n",
                       options.config_file);
    return false;
  }

  // load configuration file in memory
  const auto& content = touca::detail::load_string_file(options.config_file);

  // parse configuration file

  rapidjson::Document parsed;
  if (parsed.Parse<0>(content.c_str()).HasParseError()) {
    touca::print_error("failed to parse configuration file\n");
    return false;
  }

  // we expect content to be a json object

  if (!parsed.IsObject()) {
    touca::print_error("expected configuration file to be a json object\n");
    return false;
  }

  for (auto it = parsed.MemberBegin(); it != parsed.MemberEnd(); ++it) {
    if (!it->name.IsString()) {
      continue;
    }
    const std::string key = it->name.GetString();
    const auto& result = it->value;
    if (key == "touca") {
      if (!result.IsObject()) {
        touca::print_error(
            "field \"touca\" in configuration file has unexpected type\n");
        return false;
      }
      parse_file_option(result, "team", options.team);
      parse_file_option(result, "suite", options.suite);
      parse_file_option(result, "revision", options.revision);
      parse_file_option(result, "api-key", options.api_key);
      parse_file_option(result, "api-url", options.api_url);
      parse_file_option(result, "offline", options.offline);
      parse_file_option(result, "single-thread", options.single_thread);

      parse_file_option(result, "config-file", options.config_file);
      parse_file_option(result, "output-dir", options.output_dir);
      parse_file_option(result, "log-level", options.log_level);
      parse_file_option(result, "save-as-binary", options.save_binary);
      parse_file_option(result, "save-as-json", options.save_json);
      parse_file_option(result, "skip-logs", options.skip_logs);
      parse_file_option(result, "redirect-output", options.redirect);
      parse_file_option(result, "overwrite", options.overwrite);
      parse_file_option(result, "testcase-file", options.testcase_file);
      continue;
    }
    if (result.IsString()) {
      options.extra.emplace(key, result.GetString());
      continue;
    }
    if (result.IsBool()) {
      options.extra.emplace(key, result.GetBool() ? "true" : "false");
      continue;
    }
  }

  return true;
}

bool parse_api_url(FrameworkOptions& options) {
  // it is okay if configuration option `--api-url` is not specified
  if (options.api_url.empty()) {
    return true;
  }
  touca::ApiUrl api(options.api_url);
  if (options.team.empty() && !api._team.empty()) {
    options.team = api._team;
  }
  if (options.suite.empty() && !api._suite.empty()) {
    options.suite = api._suite;
  }
  if (options.revision.empty() && !api._revision.empty()) {
    options.revision = api._revision;
  }
  return true;
}

}  // namespace detail

std::string cli_help_description() { return detail::cli_options().help(); }

bool parse_options(int argc, char* argv[], FrameworkOptions& options) {
  auto ret = true;
  ret &= detail::parse_cli_options(argc, argv, options);
  ret &= detail::parse_file_options(options);
  parse_env_variables(options);
  ret &= detail::parse_api_url(options);
  return ret;
}

}  // namespace touca
